### PR TITLE
[#67] [Backend] As a user, I can submit my survey response

### LINKED ISSUE #32: [Android] [UI] As a user, I can see a list of surveys on the home screen
## Why

Once logged in, users can navigate through a list of surveys that they can take. The mobile app will show the basic info of the survey as they scroll through each survey.

## Acceptance Criteria

- Implement a horizontal scroll view with a snapping behavior
- Display the background image view
- Display the Page indicators
- Display the Title label
    - The maximum number of lines is 4
- Display the Description label
    - The maximum number of lines is 2
- Display the Survey Disclosure button

## Design

Image

[Home](https://www.figma.com/embed?embed_host=notion&url=https%3A%2F%2Fwww.figma.com%2Fproto%2FGjRPOjDyZ6f4EDL3wKarRK%2FChallenge-Mobile-App%3Fpage-id%3D0%253A1%26node-id%3D30%253A46%26scaling%3Dmin-zoom)

### LINKED ISSUE #67: [Backend] As a user, I can submit my survey response
## Why
At the final step, users must be able to submit their responses. The mobile app must show the Submit button instead of the Next button to inform users that the app will submit their response on the next action.

## Acceptance Criteria
Call `POST /api/v1/responses` endpoint with the following body:

```
{
  "survey_id": "{{survey-id}}",
  "questions": [
    {{question-id-with-answers}},
    ...
  ]
}
```

For `question-id-with-answers`, there are two possible types of this data structure:

1. With ID
```
{
"id": "{{question-id}}",
"answers": [
 {
   "id": "{{answer-id}}"
 }
]
}
```

2. With ID and Answer
```
{
"id": "{{question-id}}",
"answers": [
 {
   "id": "{{answer-id}}",
   "answer": "{{answer}}"
 }
]
}
```

## Resources
[Postman - Nimble Survey API](https://documenter.getpostman.com/view/11835486/T1LJjTqj#8de38274-3f9f-4b27-8122-58d7399b4c69)

### PR DESCRIPTION
**Pull request by suho**
- Close #67 

## What happened 👀

Created backend functions for submitting survey response

## Insight 📝

- I created models, a use case, and tests for submitting survey response

## Proof Of Work 📹

Screenshot 2022-12-01 at 13 20 10

### CODE REVIEW

File: shared/src/commonMain/kotlin/co/nimblehq/ic/kmm/suv/data/remote/body/SurveySubmissionApiBody.kt [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import co.nimblehq.ic.kmm.suv.domain.model.AnswerSubmission
+import co.nimblehq.ic.kmm.suv.domain.model.QuestionSubmission
+import co.nimblehq.ic.kmm.suv.domain.model.SurveySubmission
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class SurveySubmissionApiBody(
+    @SerialName("survey_id")
+    val surveyID: String,
```
Comment (luongvo) [inline shared/src/commonMain/kotlin/co/nimblehq/ic/kmm/suv/data/remote/body/SurveySubmissionApiBody.kt:12]: same

```suggestion
    val surveyId: String,
```

**Review by luongvo — COMMENTED** [review]
rest lgtm

File: shared/src/commonMain/kotlin/co/nimblehq/ic/kmm/suv/data/remote/apiclient/core/ApiClient.kt [hunk 2]
```diff
@@ -108,4 +108,15 @@ class ApiClient(
             }
         }
     }
+
+    fun emptyBody(builder: HttpRequestBuilder): Flow<Unit> {
```
[the remark this answers is not in the document]
Comment (suho) [reply shared/src/commonMain/kotlin/co/nimblehq/ic/kmm/suv/data/remote/apiclient/core/ApiClient.kt:112]: how about I change both to `reponseBody` and `emptyResponseBody`? 🤔

[the remark this answers is not in the document]
Comment (luongvo) [reply shared/src/commonMain/kotlin/co/nimblehq/ic/kmm/suv/data/remote/apiclient/core/ApiClient.kt:112]: @suho I think `reponseBody` and `emptyResponseBody` make more sense 👍

File: shared/src/commonMain/kotlin/co/nimblehq/ic/kmm/suv/domain/model/SurveySubmission.kt [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package co.nimblehq.ic.kmm.suv.domain.model
+
+data class SurveySubmission(
+    val surveyID: String,
```
[the remark this answers is not in the document]
Comment (suho) [reply shared/src/commonMain/kotlin/co/nimblehq/ic/kmm/suv/domain/model/SurveySubmission.kt]: Fixed in https://github.com/suho/kmm-ic/pull/111/commits/4ce0598b700739a8d6220950c3bfe9fb2c432212

File: shared/src/commonMain/kotlin/co/nimblehq/ic/kmm/suv/data/remote/apiclient/core/ApiClient.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (suho) [reply shared/src/commonMain/kotlin/co/nimblehq/ic/kmm/suv/data/remote/apiclient/core/ApiClient.kt]: Fixed in https://github.com/suho/kmm-ic/pull/111/commits/4ce0598b700739a8d6220950c3bfe9fb2c432212

File: shared/src/commonMain/kotlin/co/nimblehq/ic/kmm/suv/data/remote/body/SurveySubmissionApiBody.kt (hunk 1 above)
Comment (suho) [reply shared/src/commonMain/kotlin/co/nimblehq/ic/kmm/suv/data/remote/body/SurveySubmissionApiBody.kt]: Fixed in https://github.com/suho/kmm-ic/pull/111/commits/4ce0598b700739a8d6220950c3bfe9fb2c432212

File: shared/src/commonMain/kotlin/co/nimblehq/ic/kmm/suv/data/remote/body/SurveySubmissionApiBody.kt [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    @SerialName("questions")
+    val questions: List<QuestionSubmissionApiBody>
+) {
+    constructor(surveySubmission: SurveySubmission) : this(
+        surveySubmission.surveyID,
+        surveySubmission.questions.map { QuestionSubmissionApiBody(it) }
```
[the remark this answers is not in the document]
Comment (suho) [reply shared/src/commonMain/kotlin/co/nimblehq/ic/kmm/suv/data/remote/body/SurveySubmissionApiBody.kt]: That works! Nice suggestion! Fixed in https://github.com/suho/kmm-ic/pull/111/commits/4ce0598b700739a8d6220950c3bfe9fb2c432212

**Review by hoangnguyen92dn — APPROVED** [review]
Go go 🚀

**Review by luongvo — APPROVED** [review]

**Pull request merged**